### PR TITLE
Further simplify skipping of plies with threads

### PR DESCRIPTION
**Pull request by pb00068**
This is a modified version of ceebo's proposal at #574
Compared to current master:

- it requires less lines of code
- it is IMO easier to understand
- it is probably faster in execution (even if at this top-level it is not really relevant)
- it produces the identical scheme of skipping plies (verified with debug code)

no functional change

### CODE REVIEW

**Comment by vondele** [comment]
@pb00068 comparing current master with your new implementation, yours is an inlined version of master, where a while loop has been replaced by a hard-coded array (skipsize[h] and ones are the same thing), and (m/n)%2==0 has been written as m%2*n<n . The phase array is not needed (see return statement of current master), but might make the intention clearer.

Your pull request is certainly a reasonable implementation, if arrays of hard-coded numbers are to be preferred over executable code.

**Comment by pb00067** [comment]
@vondele  
expression _(m/n)%2==0_ is in fact nicer as _m%2*n<n_ because each param appears only once.
Will adapt  the pull request...

**Comment by pb00068** [comment]
@vondele
expression (m/n)%2==0 is in fact nicer as m%2*n<n because each param appears only once.
Will adapt the pull request...

**Comment by pb00068** [comment]
Pull request adapted as described in comment above without _==0_ expression to maintian the same skipping scheme (verified again with debug code).

**Comment by ddugovic** [comment]
Does anyone have recommendations for better constant names than `skipsize` and `phase`?  If I remember correctly a different file contains `GamePhase`.

**Comment by zamar** [comment]
@pb00068: I'm fine with committing this, but please provide code that you used to verify that the old and new implementation are identical. (e.g. you can push a git branch including the debug code and refer to it here...)

**Comment by pb00068** [comment]
@zamar You can find the requested branch [here](https://github.com/pb00068/Stockfish/tree/simplify_skipping_debug)
Starting the engine with no parameters should give an output like this:
```...
 idx      master                                                              patch
 1     1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0     1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0
 2     0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1     0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1
 3     0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0     0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0
 4     1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0     1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0
 5     1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1     1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1
 6     0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1     0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1
 7     0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0     0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0
 8     0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0     0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0
 9     1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0     1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0
10     1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1     1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1
11     1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1     1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1
12     0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1     0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1
13     0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0     0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0
14     0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0     0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0
15     0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0     0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0
16     1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0     1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0
17     1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1     1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1
18     1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1     1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1
19     1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1     1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1
20     0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1     0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1
21     1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0     1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0
22     0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1     0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1
23     0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0     0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0
24     1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0     1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0
```
As further check you can launch SF in selfplay with cute-chess like following:

cutechess-cli -repeat -rounds 1 -engine name=new cmd=stockfish -engine name=new cmd=stockfish -each proto=uci tc=60+0.6 option.Threads=30

Should the skipping scheme differ, then SF aborts in search.cpp at line 394

**Comment by vondele** [comment]
@pb00068 what about @ddugovic suggestion and rename the generic 'phase' into e.g. 'skipPhase'

File: src/search.cpp [hunk 1]
```diff
@@ -358,13 +348,15 @@ void Thread::search() {
 
   multiPV = std::min(multiPV, rootMoves.size());
 
+  int h = (idx - 1) % 20; // helper index, cycle after 20 threads
```
Comment (vondele) [inline src/search.cpp]: and maybe hIdx instead of h ?

Comment (pb00068) [inline src/search.cpp]: Naming variables is a matter of taste, so I will let maintainers decide which names are the best.

Comment (snicolet) [inline src/search.cpp]: `helper`could also make sense here, in my opinion:

````
int helper = (idx - 1) % 20; // helper index, cycle after 20 threads

...

if (idx && ((rootDepth / ONE_PLY + rootPos.game_ply() + phase[helper]) / skipsize[helper]) % 2)
````

**Comment by mstembera** [comment]
@pb00068  @ceebo
With this
http://tests.stockfishchess.org/tests/view/5b3906d70ebc5902b2e5fec5
machine online perhaps we could extend the skipsize and phase arrays to 40 and actually test it.  I would do it but I'm not sure what the additional values would be?

**Comment by pb00068** [comment]
@mstembera 

with highest threads skipping 5 plies at once, it will look like this:

static int skipsize[30] = {1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5};
static int phase   [30] = {0, 1, 0, 1, 2, 3, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
 ...
 int h = (idx - 1) % 30; 

I suggest to first try this one above as it requires the last amount of threads & resources.
Anyway if you aim to extend it to higher cycles, just repeat the pattern (the block-size increases linearly by 2, so for extending it with skip-size 6 the array size will be 30 + 12 = 42) 

P.S.: As such test will require an enormous  amount of resources to be meaningful, I strongly suggest to discuss and agree the test parameters (TC, bounds, hash) at the forum before submitting.
There had already been an attempt to extend the map, but at that time we hadn't the resources to test it with enough amount of games, see #727

**Comment by mstembera** [comment]
@pb00068 
Thanks!  What about this to start.. I will schedule two indication tests of 20k games @20+0.2 TC hash=512MB.  One for the skip size 30 version w/ th=30 and one for the skip size 42 version w/ th=42. (The first will finish about 2x faster than the 1k 60+0.6 th=384 scaling test running now and the second about 1.5x faster so hopefully reasonable.)  After we know the results we can better decide what(if anything) makes most sense going forward and we will have learned something.
@snicolet  @noobpwnftw Are you ok with this?

**Comment by noobpwnftw** [comment]
@mcostalba 
Of course. And the worker only accepts th>=32 because cutechess-cli cannot handle many concurrent games.

**Comment by pb00068** [comment]
In this context I like to pinpoint that the scheme looses it's efficiency with increasing thread size.
To be precise, the usefulness of the last block where we skip 4-plies at once was never tested specifically.
So maybe the first step and cheapest thing to do, is to measure it's actual benefit: 
if there is none, then further extending the scheme makes no sense.
So maybe we should test with 21 threads current master against this scheme.

static int skipsize[12] = {1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3};
static int phase [12] = {0, 1, 0, 1, 2, 3, 0, 1, 2, 3, 4, 5};
...
int h = (idx - 1) % 12;

**Comment by mstembera** [comment]
@pb00068 That makes sense.  I didn't realize 4-ply was never tested.  Test scheduled here
http://tests.stockfishchess.org/tests/view/5b3d3e940ebc5902b9ffdf06

**Comment by mstembera** [comment]
So the result is:
ELO: -0.73 +-2.8 (95%) LOS: 30.5%
Total: 20000 W: 3380 L: 3422 D: 13198
With more than 21 threads the 4-ply might be more useful.  In any case I think the skipsize array is good the way it is.

**Comment by pb00068** [comment]
> With more than 21 threads the 4-ply might be more useful

Yes, could be even if I doubt.
Maybe we should also verify with 43 threads (= TCEC-conditions). Maybe also tc=10+0.1 instead of 20+0.2. What do you think?

**Comment by mstembera** [comment]
I have no objection.  Schedule whatever test you think makes most sense.

**Pull request closed without merging**